File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int helper(vector<int>& nums, int ind, int prev, vector<vector<int>>& dp) {
        // base
        if (ind < 0)
            return 0;
        if (dp[ind][prev + 1] != -1)
            return dp[ind][prev + 1];
        int len = helper(nums, ind - 1, prev, dp);
        if (prev == -1 || nums[ind] > nums[prev]) {
            len = max(len, 1 + helper(nums, ind - 1, ind, dp));
        }
        return dp[ind][prev + 1] = len;
    }
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n + 1, vector<int>(n + 1, -1));
        reverse(nums.begin(), nums.end());
        return helper(nums, n - 1, -1, dp);
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
int helper(int[] nums, int index, int prev, int[][] dp) {
    int n = nums.length;
    if (index == n) return 0;
    if (dp[index][prev + 1] != -1) return dp[index][prev + 1];
    
    int left = 0;
    if (prev == -1 || nums[prev] < nums[index]) {
        left = 1 + helper(nums, index + 1, index, dp);
    }
    int right = helper(nums, index + 1, prev, dp);  // skip: prev unchanged
    return dp[index][prev + 1] = Math.max(left, right);
}

public int lengthOfLIS(int[] nums) {
    int n = nums.length;
    int[][] dp = new int[n][n + 1];  // prev+1 ranges 0..n
    for (int[] temp : dp) Arrays.fill(temp, -1);
    return helper(nums, 0, -1, dp);
}
}